### ISSUE TITLE
local s3 storage

### ISSUE THREAD

**Issue by bezero** [NVIDIA/tensorrt-inference-server#1023]
**Description**
I have a local s3 running and I want to set `--model-store` to the s3 bucket. However, I am not able to pass the credentials for my local s3. 

**TRTIS Information**
What version of TRTIS are you using?
`nvcr.io/nvidia/tensorrtserver:19.12-py3`
Are you using the TRTIS container or did you build it yourself?
I am using the TRTIS container

**To Reproduce**
1. Create the aws config folder
```
.aws/
   config
   credentials
```
config:
```
[default]
s3 =
    endpoint_url=http://minio:9000
    signature_version = s3v4

[plugins]
endpoint = awscli_plugin_endpoint
```
credentials:
```
[default]
aws_access_key_id='foo'
aws_secret_access_key='bar'
```
2. Create new image from TRTIS container by adding `awscli_plugin_endpoint` and `.aws` to the container. Dockerfile:
```
FROM nvcr.io/nvidia/tensorrtserver:19.12-py3

RUN pip3 install awscli_plugin_endpoint
COPY .aws /root/.aws
```
3. build `trtis` image from `Dockerfile
4. run server
```
docker run --rm trtis trtserver --model-store=s3://models
```

When I run TRTIS I get the following error:
```
E0113 20:41:25.707104 1 main.cc:1158] error: creating server: INTERNAL - Could not get MetaData for bucket with name models
```

**Expected behavior**
I expected it to use credentials and endpoint provided in the `.aws` folder and connect to my `s3 bucket`. What would be the proper way of using the local s3 storage? How to provide access key, secret and endpoints?

**Comment by bezero** [comment]
I tried to create `s3 storage` on my amazon account and run trtserver. However, I am getting the same error.

**To Reproduce:**
Steps to reproduce the behavior:
1. Create the aws config folder with awscli
2. Run:
```
docker run --rm --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 -p8000:8000 -p8001:8001 -p8002:8002 -v /home/asdf/.aws/:/root/.aws nvcr.io/nvidia/tensorrtserver:19.12-py3 trtserver --model-repository=s3://trtserver/models/
```

I am getting same error:
```
E0114 13:13:31.924762 1 main.cc:1158] error: creating server: INTERNAL - Could not get MetaData for bucket with name trtserver
```

**Comment by deadeyegoodwin** [comment]
Take a look at this test, perhaps it will help: https://github.com/NVIDIA/tensorrt-inference-server/tree/master/qa/L0_storage_S3

An empty S3 bucket needs to be available. Security credentials need to be generated for the bucket. Once we generate credentials we need to configure the client. A minimum configuration procedure is the following:
Create a file ~/.aws/credentials. The file will contain the following lines

[default]
aws_access_key_id = XXXXXX
Aws_secret_access_key = XXXXXX

Create a file  ~/.aws/config that has the following lines. Note: The region field should match the bucket’s region. 

[default]
region = XXXXX
Finally, on the test machine set the AWS_REGION variable with export AWS_REGION=XXXXX. Tensorflow’s S3 cloud storage implementation requires this variable to be set. 
For the test the aws CLI must be installed (you can install it with pip install awscli)
Open qa/L0_storage_S3/test.sh. Change the variable BUCKET_URL to point to the empty cloud storage bucket. For example, BUCKET_URL=“s3://models”.
Run qa/L0_storage_S3/test.sh. To determine if the test passes check the return code. The test will return zero if successful, non-zero if failure. The test may take several minutes to complete.

**Comment by bezero** [comment]
@deadeyegoodwin thank you for providing a detailed explanation on how to debug this issue. The reason was that I had not provided the bucket’s region in the ` ~/.aws/config` file. After adding the region, TRTIS successfully connected to amazon S3 and loaded models from there.

However, the issue with connecting to the local s3 remains. As I explained, I had tried to provide `endpoint-url` in `~/.aws/config` file using `awscli_plugin_endpoint` but it did not work. Is there a way to set `endpoint-url` using  the TRTIS container? Or should I build TRTIS from source by adding *endpoint* in the `Aws::Client::ClientConfiguration`?

**Comment by deadeyegoodwin** [comment]
Perhaps there is something with "local s3" that we need to handle differently. By local I take it that you are running your own S3-compatible server? Can you give more details, especially around what you mean about "adding endpoint in the Aws::Client::ClientConfiguration"

**Comment by bezero** [comment]
@deadeyegoodwin 
**You are running your own S3-compatible server?** Yes, I am running local S3 using [MinIO](https://min.io/)
**More details:** in `filesystem.cc` file inside `S3FileSystem::S3FileSystem` where `s3client` is created we can add the endpoint *url*. If I am not wrong, it would look something like:
```
Aws::Client::ClientConfiguration config("default");
config.endpointOverride = "localhost:<minio-port>";
client_ = s3::S3Client(config);
```
I think if it works, than `endpointOverride` can be added as an option and used as:
```trtserver --endpointOverride="localhost:<minio-port>"```

**Comment by CoderHam** [comment]
To be fixed with #1059, will close when merged into master
